common/cryptfs.c

```diff
@@ -546,47 +619,115 @@ cryptfs_write_zeros(char *crypto_blkdev, size_t size)
+		integrity = true;
+		encrypt = false;
+		break;
+	default:
+		ERROR("Unsupported type and or parameter combination");
+		return NULL;
+	}
+	/* validated parameters */
 
-	/* check if meta device is initialized */
-	initial_format = get_provided_data_sectors(meta_blkdev) != fs_size;
+	/*
+	 * Use the first 128 hex digits (64 byte) of master key for 512 bit xts mode
+	 * or full master key for AUTHENC mode
+	 */
+	crypto_key = crypto_key_len > 0 ? mem_strndup(key, crypto_key_len) : NULL;
 
-	if (create_integrity_blk_dev(real_blkdev, meta_blkdev, integrity_dev_label, fs_size) < 0) {
-		DEBUG("create_integrity_blk_dev failed!");
-		mem_free0(integrity_dev_label);
+	/* Use the following 64 hex digits (32 byte) of master key for 256 bit hmac */
+	integrity_key =
+		integrity_key_len > 0 ? mem_strndup(key + crypto_key_len, integrity_key_len) : NULL;
+
+	DEBUG("KEYS:\n\t key:\t %s\n\t split:\t %s:%s", key, crypto_key, integrity_key);
+
+	/* Update the fs_size field to be the size of the volume */
+	if ((fd = open(real_blkdev, O_RDONLY)) < 0) {
+		ERROR("Cannot open volume %s", real_blkdev);
 		goto error;
 	}
+	// BLKGETSIZE64 returns size in bytes, we require size in sectors
+	int sector_size = dm_get_blkdev_sector_size(fd);
+	if (!(sector_size > 0)) {
+		ERROR("dm_get_blkdev_sector_size returned %d\n", sector_size);
+		close(fd);
+		goto error;
+	}
+	fs_size = dm_get_blkdev_size64(fd) / sector_size;
+	close(fd);
 
-	integrity_dev = create_device_node(integrity_dev_label);
-	mem_free0(integrity_dev_label);
-
-	if (!integrity_dev) {
-		ERROR("Could not create device node");
-		return NULL;
-	} else {
-		DEBUG("Successfully created device node");
+	if (fs_size == 0) {
+		ERROR("Cannot get size of volume %s", real_blkdev);
+		goto error;
 	}
+	DEBUG("Crypto blk device size: %" PRIu64, fs_size);
 
-	if (create_crypto_blk_dev(integrity_dev, key, label, fs_size, true) < 0) {
-		ERROR("Could not create crypto block device");
-		mem_free0(integrity_dev);
-		return NULL;
+	if (integrity) {
+		integrity_dev_label = mem_printf("%s-%s", label, "integrity");
+		TRACE("cryptfs_setup_volume_integrity_new");
+
+		/* check if meta device is initialized */
+		initial_format = get_provided_data_sectors(meta_blkdev) != fs_size;
+
+		if (!(integrity_blkdev =
+			      create_integrity_blk_dev(real_blkdev, meta_blkdev, integrity_key,
+						       integrity_dev_label, fs_size, stacked))) {
+			DEBUG("create_integrity_blk_dev failed!");
+			goto error;
+		}
 	}
 
-	crypto_blkdev = create_device_node(label);
-	if (!crypto_blkdev) {
-		ERROR("Could not create device node");
-		mem_free0(integrity_dev);
-		return NULL;
+	if (encrypt) {
+		if (!(crypto_blkdev = create_crypto_blk_dev(meta_blkdev ? integrity_blkdev : real_blkdev,
+							    crypto_key, label, fs_size, stacked))) {
+			ERROR("Could not create crypto block device");
+			goto error;
+		}
 	} else {
-		DEBUG("Successfully created device node");
+		crypto_blkdev = integrity_blkdev;
```

Comment:
is if(integrity) missing here?
(if parameters are integrity=false and encrypt=true)